pandas/core/internals/concat.py

```diff
@@ -31,6 +31,47 @@
     from pandas import Index
 
 
+def concatenate_array_managers(
+    mgrs_indexers, axes: List[Index], concat_axis: int, copy: bool
+) -> Manager:
+    """
+    Concatenate array managers into one.
+
+    Parameters
+    ----------
+    mgrs_indexers : list of (ArrayManager, {axis: indexer,...}) tuples
+    axes : list of Index
+    concat_axis : int
+    copy : bool
+
+    Returns
+    -------
+    ArrayManager
+    """
+    # reindex all arrays
+    mgrs = []
+    for mgr, indexers in mgrs_indexers:
+        for ax, indexer in indexers.items():
+            mgr = mgr.reindex_indexer(
+                axes[ax], indexer, axis=ax, allow_dups=True, do_integrity_check=False
+            )
+        mgrs.append(mgr)
+
+    # concatting along the rows -> concat the reindexed arrays
+    # TODO(ArrayManager) doesn't yet preserve the correct dtype
+    if concat_axis == 1:
+        arrays = [
+            concat_compat([mgrs[i].arrays[j] for i in range(len(mgrs))])
+            for j in range(len(mgrs[0].arrays))
+        ]
+        return ArrayManager(arrays, [axes[1], axes[0]], do_integrity_check=False)
+    # concatting along the columns -> combine reindexed arrays in a single manager
+    else:
+        assert concat_axis == 0
+        arrays = list(itertools.chain.from_iterable([mgr.arrays for mgr in mgrs]))
+        return ArrayManager(arrays, [axes[1], axes[0]], do_integrity_check=False)
+
+
 def concatenate_block_managers(
```

Comment:
can you at a higher level call concatenate_array_managers instead? (e.g. instead of dispatching thru the BM)

Reply:
That would require adding if/else checking the manager type in each of the places where currently `concatenate_array_managers` is called (3 places), so that would not improve the situation.

If you want I can create a new `concatenate_managers` in this file that simply does the dispatch to `concatenate_block_managers` / `concatenate_array_managers`, and call that instead of `concatenate_block_managers` outside of the `/internals`

Reply:
ok maybe rename this then to concatenate_managers? to make it much more obvious.